### ISSUE TITLE
OUT_OF_RANGE while invoking hal.fence.join for non-tiny model training

### ISSUE THREAD

**Issue by phoenix-meadowlark** [iree-org/iree#10903]
### What happened?

`iree-run-module` throws an [`OUT_OF_RANGE` error](https://github.com/iree-org/iree/blob/main/runtime/src/iree/modules/hal/module.c#L1037-L1038) when executing the training step for models which train more than ~32 variables. For example, when attempting to execute training for a model with 36 variables the following error is thrown:

```
iree/runtime/src/iree/modules/hal/module.c:1038:
  OUT_OF_RANGE; count 37 of iree_hal_fence > 32; while invoking native function hal.fence.join; while calling import;
```

Merely updating more than ~32 variables is insufficient so I've included the whole training step in the reproducer. The count above also doesn't solely depend on the number of variables.

### Steps to reproduce your issue

The error above can be reproduced using the following reproducer: [`out_of_range.mlir`](https://gist.github.com/phoenix-meadowlark/ba0f76235510dc3c01ff0e5aba7d0f94)

```shell
iree-compile \
  --iree-hal-target-backends=vmvx \
  --iree-input-type=mhlo \
  /tmp/out_of_range.mlir \
  -o /tmp/out_of_range.vmfb

iree-run-module \
  --device=local-sync \
  --entry_function=train_step \
  --function_input=100x16000xi16 \
  --function_input=100x16000xf32 \
  --function_input=100xi16 \
  --module_file=/tmp/out_of_range.vmfb
```

```
EXEC @train_step
iree/runtime/src/iree/modules/hal/module.c:1038: OUT_OF_RANGE; count 37 of iree_hal_fence > 32; while invoking native function hal.fence.join; while calling import;
[ 1]   native hal.fence.join:0 -
[ 0] bytecode hotword.train_step:4038 [
  ]
      at /tmp/out_of_range.mlir:75:14
      at /tmp/out_of_range.mlir:38:3; invoking function 'train_step'
```

### What component(s) does this issue relate to?

Runtime

### Version information

cfedd317979496241f4db26343f358b9f1845879

**Comment by benvanik** [comment]
Interesting that these don't get coalesced - I suspect maybe the missing IPO step across an internal function call boundary. Will take a look.

**Comment by benvanik** [comment]
(interesting stuff in this model - lots of opportunities for dispatch formation improvements - seeing a lot of tensor<f32>->tensor<f32> dispatches that could be merged)

**Comment by benvanik** [comment]
Also some shady things:
`stream.cmd.dispatch @jit__train_step$main_dispatch_48::@jit__train_step$main_dispatch_48_generic(%c0, %c64, %c128 : index, index, index)`
That's a no-op.

And yeah, this fusion stuff is real bad:
![image](https://user-images.githubusercontent.com/75337/197892931-273ae49c-8ca1-4078-83bb-046e12a3021b.png)

Whenever someone looks at performance for this model this should definitely be the starting point!

**Comment by benvanik** [comment]
Confirmed this is caused by lack of IPO; few of these around where every single one of those args is the same value in the caller:
```
%3 = stream.timepoint.join max(%arg228, %arg18, %_constant__timepoint, %arg54, %arg48, %arg66, %arg60, %arg30, %arg24, %arg42, %arg36, %arg78, %arg72, %arg96, %arg102, %arg162, %arg108, %arg168, %arg114, %arg174, %arg120, %arg180, %arg126, %arg186, %arg132, %arg192, %arg138, %arg198, %arg144, %arg204, %arg150, %arg210, %arg156, %arg216, %arg84, %arg90, %arg222, %result_timepoint) => !stream.timepoint
```

Will need to implement that pass to make this work.

**Comment by benvanik** [comment]
I've made progress here and have the fence issue fixed - I've got to write tests tomorrow but you can check it out here while I do that: #10911.

I'm getting assertion failures in CPU codegen doing an arith::CmpIOp folder on vector types. That's terrifying but unrelated. I suspect you're running a release compiler build with assertions disabled otherwise you'd have hit that long before you hit runtime issues :)

btw just remembered you can use `--mlir-print-debuginfo=false` to hide all those source location lines from your errors if they're making things tricky for you.

**Comment by phoenix-meadowlark** [comment]
> I suspect you're running a release compiler build with assertions disabled otherwise you'd have hit that long before you hit runtime issues :)

I've been using `RelWithDebInfo` – would `Debug` be a better option when just trying to test what is and isn't supported?

**Comment by ScottTodd** [comment]
> > I suspect you're running a release compiler build with assertions disabled otherwise you'd have hit that long before you hit runtime issues :)
> 
> I've been using `RelWithDebInfo` – would `Debug` be a better option when just trying to test what is and isn't supported?

https://iree-org.github.io/iree/building-from-source/getting-started/#clone-and-build

```
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DIREE_ENABLE_ASSERTIONS=ON \
```

**Comment by phoenix-meadowlark** [comment]
> > > I suspect you're running a release compiler build with assertions disabled otherwise you'd have hit that long before you hit runtime issues :)
> > 
> > 
> > I've been using `RelWithDebInfo` – would `Debug` be a better option when just trying to test what is and isn't supported?
> 
> https://iree-org.github.io/iree/building-from-source/getting-started/#clone-and-build
> 
> ```
>     -DCMAKE_BUILD_TYPE=RelWithDebInfo \
>     -DIREE_ENABLE_ASSERTIONS=ON \
> ```

Okay yeah that's what I've been using!

**Comment by benvanik** [comment]
Interesting - I wonder why you don't hit the same assert I am in #10915 - may be worth trying one time in Debug just to make sure the IREE_ENABLE_ASSERTIONS flag is working as it should (I only run in Debug).

**Comment by phoenix-meadowlark** [comment]
Hmmm, I did a `Debug` build from scratch with `IREE_ENABLE_ASSERTIONS` unset and I'm not hitting any asserts when I compile [`out_of_range.mlir`](https://gist.github.com/phoenix-meadowlark/ba0f76235510dc3c01ff0e5aba7d0f94).

In both configurations trying to execute the VMVX-compiled module causes `iree-run-module` to hang after printing `EXEC @train_step`. I actually noticed this behavior for some other modules as well, but wasn't sure if it was a bug or VMVX just being incredibly slow for some reason.

**Comment by benvanik** [comment]
VMVX is incredibly slow - try llvm-cpu :)

**Comment by phoenix-meadowlark** [comment]
Gotcha! Training with llvm-cpu is currently blocked by https://github.com/iree-org/iree/issues/10839 so I switched to VMVX to see if I could find any other that might pop up down the line.

**Comment by benvanik** [comment]
Oh nice - that's a duplicate of the issue I was hitting - we should make sure @MaheshRavishankar / @hanhanW / someone codegen sees it!
